### PR TITLE
Game history

### PR DESCRIPTION
**Pull request by vitsalis**

### CODE REVIEW

File: tests/models/error.php [hunk 1]
```diff
@@ -18,6 +20,8 @@ public function testCreate() {
             $this->assertEquals( 1, count( $errorObjects ), 'The number of errors in the database must be the same as the ones created' );
 
             $this->assertEquals( 'error', $errorObjects[ 0 ]->error, 'The error in the database must be the one during creation' );
+            $this->assertEquals( 'actual', $errorObjects[ 0 ]->actual, 'The actual in the database must be the one during creation' );
+            $this->assertEquals( 'actual', $errorObjects[ 0 ]->actual, 'The actual in the database must be the one during creation' );
```
Comment (dionyziz) [inline tests/models/error.php]: expected

File: tests/bots/php/bot.php [hunk 2]
```diff
@@ -187,6 +187,60 @@ public function testSendRoundRequestNoNeighbors() {
...
+                        'hp' => 0
+                    ]
+                ] ),
+                'gameid' => 1,
+                'myid' => 1,
+                'W' => 100,
+                'H' => 100
+            ] );
+
+            $this->assertTrue( $response->intent[ 0 ]->action !== 'ATTACK', 'A creature must not try and attack a dead creature' );
```
Comment (dionyziz) [inline tests/bots/php/bot.php:215]: dead-horse

File: models/game.php [hunk 3]
```diff
@@ -232,6 +232,11 @@ public function nextRound() {
             }
             $this->rounds[ $roundid ]->save();
         }
+        public function beforeNextRound() {
+            foreach ( $this->getCurrentRound()->creatures as $creature ) {
+                $creature->intent = new Intent();
+            }
+        }
```
Comment (dionyziz) [inline models/game.php:239]: This seems like a serious bug; regression test that intents are cleared in every new round.

File: controllers/game.php [hunk 4]
```diff
@@ -18,19 +18,27 @@ public function createView() {
         }
         public function update( $gameid ) {
             $game = new Game( $gameid );
+            $game->beforeNextRound();
```
Comment (dionyziz) [inline controllers/game.php]: Don't call this from here. Call it from `$grader->nextRound();`.

File: models/game.php [hunk 5]
```diff
@@ -232,6 +232,11 @@ public function nextRound() {
             }
             $this->rounds[ $roundid ]->save();
         }
+        public function beforeNextRound() {
```
Comment (dionyziz) [inline models/game.php:235]: I suggest that you don't modify the intents of the previous round. This is semantically incorrect. Recommend that you create a new round first, then clean up past intents, then request intents from bots and fill them in.

File: models/grader/bot.php [hunk 6]
```diff
@@ -110,21 +110,25 @@ public function sendGameRequest( Game $game ) {
...
-                $ch = $this->httpRequest( "round", 'create', GraderSerializer::roundRequestParams( $round, $this->user ) );
+                $ch = $this->httpRequest( "round", 'create', GraderSerializer::roundRequestParams( $round, $this->user, $this->game ) );
             }
             catch ( CurlException $e ) {
                 $this->reportError( $e->error );
             }
             $decodedResponse = json_decode( $ch->response );
             if ( $decodedResponse === null ) {
-                $this->reportError( 'round_invalid_json' );
+                $this->reportError( 'round_invalid_json', '', $ch->response );
```
Comment (dionyziz) [inline models/grader/bot.php:131]: Test that actual is reported correctly for invalid JSON.

File: views/game/view.php [hunk 7]
```diff
@@ -1,7 +1,28 @@
...
+                ?>px; position: absolute; background-color: <?php
+                if ( !isset( $playerColor[ $creature->user->id ] ) ) {
+                    $playerColor[ $creature->user->id ] = array_shift( $colors );
+                }
+                echo $playerColor[ $creature->user->id ];
+                ?>; width: 25px; height: 25px;"></div><?php
+            }
+        }
+        ?>
+    </div><?php
```
Comment (dionyziz) [inline views/game/view.php:27]: Take a screenshot of this and put it on the blog, it's pretty cool.

File: tests/bots/php/bot.php (hunk 2 above)
Comment (vitsalis) [inline tests/bots/php/bot.php:215]: HAHAHAHA

File: models/game.php (hunk 5 above)
Comment (vitsalis) [inline models/game.php:235]: the problem with that, is that we set the intents for the new round on $creature->intent of the previous round. So for example the intents of $game->rounds[ 0 ]->creatures are the moves that the creatures wish to do on round 1.

Comment (vitsalis) [inline models/game.php:235]: so when we are clearing the previous round's intents we just make space for our new intents. 
so round 0 has the intents that the user wishes to do oon round 1 and so on.

Comment (dionyziz) [inline models/game.php:235]: OK then

**Comment by dionyziz** [comment]
Please fix the build.

**Pull request merged**